src/main/java/project/bookstore/service/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,147 @@
+                cartItemDto.setQuantity(cartItemDto
+                        .getQuantity() + shoppingCart
+                        .getCartItems()
+                        .stream()
+                        .filter(cartItem -> cartItem.getBook().getId()
+                                .equals(cartItemDto.getBookId()))
+                        .findFirst()
+                        .get()
+                        .getQuantity());
+                addCartItemToShoppingCart(cartItemDto, book, shoppingCart);
+
+                return shoppingCartMapper.toDto(shoppingCart);
+            }
+        }
+        addCartItemToShoppingCart(cartItemDto, book, shoppingCart);
+
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Transactional
+    @Override
+    public ShoppingCartDto updateItemInShoppingCart(
+            Long cartItemId,
+            UpdateCartItemDto updateCartItemDto,
+            String username
+    ) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserEmail(username)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Shopping cart not found with username: " + username));
+        CartItem cartItem = getCartItem(cartItemId, shoppingCart.getId(), username);
+        cartItem.setQuantity(updateCartItemDto.getQuantity());
+        cartItemRepository.save(cartItem);
+        shoppingCartRepository.save(shoppingCart);
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public void deleteItemFromShoppingCart(Long cartItemId) {
+        hasAccessToShoppingCart(cartItemId);
+        cartItemRepository.deleteById(cartItemId);
+    }
+
+    @Override
+    public void addShoppingCart(User user) {
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(user);
+        shoppingCartRepository.save(shoppingCart);
+    }
+
+    private void hasAccessToShoppingCart(Long cartItemId) {
+        String authenticatedUsername = SecurityContextHolder
+                .getContext().getAuthentication().getName();
+        String username = cartItemRepository
+                .findById(cartItemId)
+                .orElseThrow(() -> new EntityNotFoundException("User not found"))
+                .getShoppingCart()
+                .getUser()
+                .getUsername();
+        if (!authenticatedUsername.equals(username)) {
+            throw new AccessDeniedException("You are not authorized to delete this item.");
+        }
+    }
+
+    private void addCartItemToShoppingCart(
+            CreateCartItemDto cartItemDto,
+            Book book,
+            ShoppingCart shoppingCart
+    ) {
+        CartItem cartItem = cartItemMapper.toModel(cartItemDto);
+        cartItem.setBook(book);
+        cartItem.setShoppingCart(shoppingCart);
+        cartItemRepository.save(cartItem);
+        shoppingCart.getCartItems().add(cartItem);
+        shoppingCartRepository.save(shoppingCart);
```

Comment:
```suggestion
        shoppingCart.addItemToCart(cartItem);
```

Where in ShoppingCart

```
public void addItemToCart(CartItem cartItem) {
        cartItem.setShoppingCart(this);
        cartItems.add(cartItem);
    }
```